### ISSUE TITLE
Reduce Envoy DNS Requests for STRICT_DNS clusters

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#13710]
## Problem

For STRICT_DNS clusters, Envoy will send a DNS requests every X seconds (5 by default). DNS lookups in Kubernetes can potentionally involve many requests. This is due to the search and ndots (default 5 in Kubernetes) configuration.

From my findings, the following number of DNS requests are needed

| Request                                                           | Lookups | Reason                                                                                            |
| ----------------------------------------------------------------- | ------- | ------------------------------------------------------------------------------------------------- |
| google.com                                                        | 6       | Looks though google.com.svc.cluster.local, etc before `google.com.`                               |
| google.com.                                                       | 1       | FQDN, so doesn't search                                                                           |
| istio-pilot.istio-system                                          | 2       | Searches `istio-pilot.istio-system.default.svc.cluster.local` (default is namespace of pod) first |
| istio-pilot.istio-system.svc.cluster.local                        | 6       | Must go through all searches first                                                                |
| istio-pilot.istio-system.svc.cluster.local.                       | 1       | FQDN, so doesn't search                                                                           |
| istio-zipkins.istio-system (without zipkins pod)                  | 6       | Searches all, NXDOMAIN for all                                                                    |
| istio-zipkin.istio-system.svc.cluster.local. (without zipkins pod) | 1       | Searches just the FQDN, NXDOMAIN                                                                  |

In a default installation, we set up clusters for `istio-pilot.istio-system` and `istio-zipkins.istio-system` (which does not exist by default). This means we will get `2 + 6` requests every 5 seconds per pod, or 1.6 RPS/Envoy. Every additional service entry will add another 1.2 RPS for each host AND each port in the service entry, per Envoy. Consider a cluster with 1000 pods, and 10 ServiceEntries with 1 host and 1 port each - we would expect 13.6K RPS just to the DNS server, which is very high.

Luckily, these DNS requests are performed asynchronously, so this is unlikely to have too much of a negative impact on latency.

The 5s lookup period is configurable, but currently this does not effect the bootstrap config (zipkins and pilot)

## Mitigation

1. Unambiguous FQDN (**with the trailing** `.`), are considering invalid by Galley and Pilot. This blocks us from using `istio-pilot.istio-system.svc.cluster.local.` as our address and reducing our DNS requests from 2 to 1. This also prevents users from using unambiguous FQDN in their ServiceEntries, which has an even bigger impact - 6 calls down to 2.
2. (If we do #1) we should use unambiguous FQDNs for the entries we add in the bootstrap configuration.
3. We should make the configurable DNS lookup time effect the bootstrap config.
4. In all our documentation we recommend using resolution: DNS on ServiceEntry to expose external services. This means that when doing curl google.com, curl resolves google.com and sends to the request to Envoy which uses its own DNS resolution, so we have double resolution going on. We should reduce examples of DNS resolution to only when it is needed, and recommend NONE instead (unless there are reasons to use DNS I am missing).
5. We should raise the default resolution interval from 5s to Xs (15s? 30s?).
6. We waste a lot of DNS requests on Zipkins in the out of the box. We should consider removing it from the bootstrap config unless it is explicitly enabled. A downside here is that pods would need restarts if the config is changed. However, this restriction already applies to switching to Datadog or Lightstep. It won't be present if we pass `global.proxy.tracer=""` rather than the default (zipkins). If we want to keep the default we should probably add a guard against `tracing.enabled`, or maybe `global.proxy.enableTracing`?

With these changes we can get the number of requests down to 1, from 8 (no zipkins) or 2, from 4 (with zipkins).

Still left unsolved is the issue of duplicating DNS lookups for each port. This would likely need to be a change in Envoy (to cache the lookups across clusters), which is probably not trivial/something that is even desired. Given most ServiceEntries seem to have 1 or 2 ports, this probably isn't too much of an issue.

After applying 2 and 6 locally, I saw DNS requests drop from, about 5x lower.

## Related Issues

https://github.com/istio/istio/issues/12968
https://github.com/istio/istio/issues/12181

**Comment by howardjohn** [comment]
Some more follow up after talking with @costinm and @fpesce.

In all our [documentation](https://istio.io/docs/tasks/traffic-management/egress/#access-an-external-https-service) we recommend using `resolution: DNS` on ServiceEntry to expose external services. This means that when doing `curl google.com`, curl resolves `google.com` and sends to the request to Envoy which uses its own DNS resolution, so we have double resolution going on. There are some cases for this (direct some host to another host for example?), but for the typical case it seems like `resolution: NONE` is a much better option. As mentioned above, using DNS resolution comes at the cost of 1.2 RPS per ServiceEntry per Envoy with default settings.

**Comment by ceastman-ibm** [comment]
For my IBM cloud cluster, i switched from kube-dns to coredns and that cleared up the issues i was having. From what I have read coredns is much better at external dns lookups than kube-dns.

**Comment by Stono** [comment]
Problem is a lot of people using GKE, and managed istio on gke just can't do that 🤷‍♂

**Comment by Stono** [comment]
Please do number 4, most people will surely use zipkin/tracing as its a big value add of istio.

**Comment by howardjohn** [comment]
Action items from networking WG:
* 5 min default
* bootstrap respects dns
* use "resolution: none" everywhere
* allow unambiguous fqdns
* open envoy issue

**Comment by Stono** [comment]
All sounds awesome

**Comment by howardjohn** [comment]
On the FQDNs: There are some complications regarding hosts.

If we do `curl google.com.` it will send `Host: google.com`. Other clients may send `Host: google.com.`. 
From our side, we could allow both variants, or pick one or the other, but its more complexity without too much gain/

Some servers will accept `Host: google.com` and `Host: google.com.`, some will 301 redirect with a `.`, etc, so the other side is complicated too.

Because we just dropped DNS requests by 60x in 1.2, I don't think we need to invest too much effort in allowing the FQDNs right now. If the DNS problem comes up again in the future then we can decide at that point.

**Comment by howardjohn** [comment]
Actually decided to support FQDN as there are some cases where users REQUIRE it due to client libraries using them which they have no control over (google metadata is one I have seen?). Added in https://github.com/istio/istio/pull/14405. 

So the only step left is https://github.com/envoyproxy/envoy/issues/6876

@crazyxy are you still working on the DNS TTL stuff?

**Comment by crazyxy** [comment]
Yes. This is blocked by the upstream c-ares. c-ares doesn't support DNS query with TTL. https://github.com/c-ares/c-ares/issues/259 Waiting for upstream change.

**Comment by crazyxy** [comment]
Respect DNS ttl has been merged into upstream Envoy. `respect_dns_ttl` is the switch to turn on/off and feature.

**Comment by howardjohn** [comment]
Nice!

**Comment by cagataygurturk** [comment]
@howardjohn 

A quick hands-up: On our GKE cluster running 1.4.6 we have 116 proxies and 61 service entries. As soon we we set all those entries' resolution to DNS, kube-dns starts complain:

`Maximum number of concurrent DNS queries reached (max: 150)"

Setting resolution to NONE fixes it.

dnsRefreshRate is set to 300s

I'd like to add that there is no too much traffic to these endpoints. They are mostly idle.

**Comment by howardjohn** [comment]
@cagataygurturk by my calculation thats about 23 QPS. Seems like there is either slow queries for some reason, thundering herd, or my calculation is not correct. Note that 300s is no longer as relevant, we now respect the TTL returned in the DNS respones, so if its very low it could impcat as well.

Traffic to the endpoints does not matter, envoy resolves async

**Comment by cagataygurturk** [comment]
@howardjohn we've scanned all the service entries and found out one domain that does not exist anymore. It is used in following TLS origination configuration:

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: example
  namespace: istio-system
spec:
  hosts:
    - not_existing_domain.com
  ports:
    - number: 80
      name: http-port
      protocol: HTTP
    - number: 443
      name: http-port-for-tls-origination
      protocol: HTTPS
  resolution: NONE
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: example
  namespace: istio-system
spec:
  hosts:
    - not_existing_domain.com
  http:
    - match:
        - port: 80
      route:
        - destination:
            host: not_existing_domain.com
            subset: tls-origination
            port:
              number: 443
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: example
  namespace: istio-system
spec:
  host: not_existing_domain.com
  subsets:
    - name: tls-origination
      trafficPolicy:
        loadBalancer:
          simple: ROUND_ROBIN
        portLevelSettings:
          - port:
              number: 443
            tls:
              mode: SIMPLE
```

Some clients try to connect to http://not_existing_domain.com still. And this increases the load on kube-dns.

Two things reduces back the load:

- Switching to resolution: NONE in ServiceEntry
- Deleting the DestinationRule, meaning that disabling TLS origination. Deleting only VirtualService leaving DestinationRule does not fix the issue.

It is important to note that this domain is still being tried by applications. If the domain is replaced by one  not used by any application, the issue is not observed.

Is this behaviour expected?

**Comment by howardjohn** [comment]
Clients connecting to the domain has no impact on envoy. With 0 RPS or 10000000 RPS it still makes the same dns requests - its all done asynchronously not in response to traffic. Any extra DNS requests to this domain from load are coming from the app (in theory at least)

**Comment by cagataygurturk** [comment]
> Any extra DNS requests to this domain from load are coming from the app (in theory at least)

I understand how Envoy works for DNS but I have tried many combinations and as soon as the `DestinationRule` above is there (to enable TLS origination) for a not existing host, DNS requests reach to peak.

Deploying **only** the simple ServiceEntry does not cause any issue under the same request volume from applications:

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: example
  namespace: istio-system
spec:
  hosts:
    - not_existing_domain.com
  ports:
    - number: 80
      name: http-port
      protocol: HTTP
  resolution: DNS
```

I wonder what changes here with the `DestinationRule`.

**Comment by cagataygurturk** [comment]
I am gonna investigate a bit more to make sure whether this is reproducible in another application environment @howardjohn, thanks for the explanations.
